### PR TITLE
Support variable coefficients in ADMatchedValueBC and MatchedValueBC

### PR DESCRIPTION
**Pull request by fdkong**
Closes #19012

### CODE REVIEW

**Comment by lindsayad** [comment]
You decided not to combine them in one class?

**Comment by fdkong** [comment]
> You decided not to combine them in one class?

I was too lazy. This change is good for the project at this moment.

File: framework/src/bcs/MatchedValueBC.C [hunk 1]
```diff
@@ -18,27 +18,33 @@ MatchedValueBC::validParams()
...
-  return _u[_qp] - _v[_qp];
+  return _u_coeff * _u[_qp] - _v_coeff * _v[_qp];
 }
 
 Real
 MatchedValueBC::computeQpOffDiagJacobian(unsigned int jvar)
 {
   if (jvar == _v_num)
-    return -1.0;
+    return -1.0 * _v_coeff;
```
Comment (lindsayad) [inline framework/src/bcs/MatchedValueBC.C:47]:
```suggestion
    return -_v_coeff;
```

File: test/tests/interfacekernels/hybrid/tests [hunk 2]
```diff
@@ -0,0 +1,37 @@
+[Tests]
+  issues = "#7437"
+   design = "InterfaceKernels/index.md"
+
+  [./matchedvalue]
+    type = 'Exodiff'
+    input = 'interface.i'
+    exodiff = 'interface_out.e'
+    requirement = "The system shall support interface conditions: Dirichlet and Neumann"
```
Comment (lindsayad) [inline test/tests/interfacekernels/hybrid/tests:9]:
```suggestion
    requirement = "The system shall support Dirichlet and Neumann interface conditions."
```

File: test/tests/interfacekernels/hybrid/tests [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  [../]
+
+  [./penalty]
+    type = 'Exodiff'
+    input = 'interface.i'
+    exodiff = 'interface_penalty.e'
+    cli_args = 'InterfaceKernels/active="penalty" BCs/active="u" Outputs/file_base=interface_penalty'
+    prereq = 'matchedvalue'
+    requirement = "The system shall support interface conditions: penalty"
```
Comment (lindsayad) [inline test/tests/interfacekernels/hybrid/tests:18]:
```suggestion
    requirement = "The system shall support penalty interface conditions."
```

File: test/tests/interfacekernels/hybrid/tests [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  [../]
+
+  [./variable_coeffs_AD]
+    type = 'Exodiff'
+    input = 'interface.i'
+    exodiff = 'variable_coeffs.e'
+    cli_args = 'BCs/interface_bc/u_coeff=1 BCs/interface_bc/v_coeff=1.08 Outputs/file_base=variable_coeffs'
+    requirement = "The system shall support matched-value interface with different variable coefficients with AD"
```
Comment (lindsayad) [inline test/tests/interfacekernels/hybrid/tests:26]:
```suggestion
    requirement = "The system shall support matched-value interface with different variable coefficients with automatic differentiation."
```

File: test/tests/interfacekernels/hybrid/tests [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  [../]
+
+  [./variable_coeffs]
+    type = 'Exodiff'
+    input = 'interface.i'
+    exodiff = 'variable_coeffs.e'
+    prereq = 'variable_coeffs_AD'
+    cli_args = 'BCs/interface_bc/type=MatchedValueBC BCs/interface_bc/u_coeff=1 BCs/interface_bc/v_coeff=1.08 Outputs/file_base=variable_coeffs'
+    requirement = "The system shall support matched-value interface with different variable coefficients"
```
Comment (lindsayad) [inline test/tests/interfacekernels/hybrid/tests:35]:
```suggestion
    requirement = "The system shall support matched-value interface with different variable coefficients."
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
LGTM

**Review by lindsayad — APPROVED** [review]

**Pull request merged**